README.md
# ant-creator
This app creates new PHP-Ant apps. It requries:

1. You have generated your signing keys with `publish -k` [See: ant-signer project on Github](https://github.com/mjmunger/ant-signer).
1. 


CHANGELOG
# v0.1.3

- Fixed newline bug.

# v0.1.2

- Fixed a bug where using "." as the specified app directory fails when compiled to a phar.
- Added this changelog.

tests/AntAppCreatorTest.php
<?php

namespace AntAppCreator;

use PHPUnit\Framework\TestCase;

/**
 * {CLASS SUMMARY}
 *
 * Date: 2/21/18
 * Time: 6:54 AM
 * @author Michael Munger <michael@highpoweredhelp.com>
 */

class AntAppCreatorTest extends TestCase
{
    public static function setUpBeforeClass() {
        $dependencies = [];
        $dependencies[] = 'classes/AppCreator.class.php';


        foreach($dependencies as $dep) {
            require_once($dep);
        }
    }
    /**
     * Test the cloning ability to create a base template from a remote repo.
     * Tests:
     * 1. Can I clone into a target directory?
     * 2. Can I read the resulting files?
     * @covers getTemplate
     */

    public function testCloneTemplate() {
        $targetDir = '/tmp/_phpant_template/';
        $Creator = new AppCreator(".");
        $Creator->getTemplate($targetDir);

        $targetFiles = [];
        $targetFiles[] = $targetDir . 'ant-app-template.php';
        $targetFiles[] = $targetDir . 'autoloader-template.php';

        foreach($targetFiles as $target) {
            $this->assertFileExists($target, "$target is missing, and should have been cloned from the repo.");
            $this->assertIsReadable($target);
        }
    }

    /**
     * @dataProvider providerAppData
     * @covers findAndReplace
     * @depends testCloneTemplate
     */

    public function testSubstitutions($data , $expectedCounts) {
        $targetDir = '/tmp/_phpant_template/';
        $Creator = new AppCreator(".");
        $Creator->getTemplate($targetDir);
        $Creator->findAndReplace($data);

        foreach($data as $field => $value) {
            $this->assertContains($value,$Creator->fileBuffer);
        }

        foreach($expectedCounts as $field => $count) {
            $this->assertSame( $count
                             , substr_count($Creator->fileBuffer,$data[$field])
                             , sprintf("Find and replace fail. The string %s should have occured %s times in the resulting fileBuffer %s"
                                      , $data[$field]
                                      , $count
                                      , $Creator->fileBuffer
                                      )
                             );
        }
    }

    /**
     * @covers verifyPrerequisites
     */

    public function testVerifyPrerequisites() {
        $Creator = new AppCreator(".");
        $this->assertTrue($Creator->verifyPrerequisites());
    }

    public function providerAppData() {

        $data = [];
        $data['PROJECT']        = 'AAvYisqiEBqlGvUPZBv';
        $data['SUBSPACE']       = 'MON';
        $data['FRIENDLYNAME']   = 'DKPGtjIcOwbs';
        $data['APPDESCRIPTION'] = 'JbYwLaHuYCFvOgS';
        $data['SYSTEMNAME']     = 'Xsi';
        $data['CATEGORY']       = 'CbbkTweJiZr';
        $data['AUTHORNAME']     = 'ZLnKkfWeY';
        $data['AUTHOREMAIL']    = 'fIzIntm';

        $expectedCounts = [];
        $expectedCounts['PROJECT']        = 2;
        $expectedCounts['SUBSPACE']       = 2;
        $expectedCounts['FRIENDLYNAME']   = 4;
        $expectedCounts['APPDESCRIPTION'] = 1;
        $expectedCounts['SYSTEMNAME']     = 10;
        $expectedCounts['CATEGORY']       = 1;
        $expectedCounts['AUTHORNAME']     = 4;
        $expectedCounts['AUTHOREMAIL']    = 3;

        return  [ [ $data , $expectedCounts] ];
    }

    /**
     * @dataProvider providerRepoURIs
     * @covers parseTargetDirName
     */

    public function testParseDirname($repoURI, $expectedDirName) {
        $Creator = new AppCreator(".");
        $targetDirName = $Creator->parseTargetDirName($repoURI);
        $this->assertSame($expectedDirName, $targetDirName);
    }

    public function providerRepoURIs() {
        return  [ [ 'git@git.highpoweredhelp.com:michael/bugzy-work-timer.git', 'bugzy-work-timer' ] ];
    }
}

install.sh
#!/bin/bash
TARGET=/usr/local/bin/create-app
PHARDIR=/usr/share/antcreator
PHARPATH=$PHARDIR/antcreator.phar
LINK=/usr/local/bin/create-app

check_root() {
if [ "$EUID" -ne 0 ]
  then echo "Please run as root"
  exit 1
fi    
}

install_files() {

    echo ""
    echo "Installing files..."

    if [ ! -f antcreator.phar ]; then
        echo "antcreator.phar is not here. Re-building phar..."
        php buildphar.php
    fi

    if [ ! -d  /usr/share/antcreator/ ]; then
        mkdir /usr/share/antcreator/
    fi

    chmod +x antcreator.phar

    
    mv antcreator.phar $PHARPATH

    echo "antcreator.phar installed to: $PHARPATH"

    if [ ! -L /usr/local/bin/create-app ]; then
        ln -s $PHARPATH $LINK
    fi

    echo "Symlink installed $LINK -> $PHARPATH"
    echo "done"
    echo ""
}

remove_files() {
    
    [ ! -L $TARGET ] && unlink $TARGET
    echo "Unlnked: $TARGET"
    rm -v $PHARPATH
    echo "Removed $PHARPATH"
    echo "Remove $PHARDIR? [y/N]"
    read CHOICE

    [ "$CHOICE" == "Y" ] && rm -vfr $PHARDIR

    echo "Uninstall complete."
}

show_help() {
    cat <<'EOF'

SUMMARY:

Installs the PHP-Ant app creator scripts.

SYNTAX:

install.sh [ install | remove ]

WHERE:
  install  Installs the scripts into /usr/local/bin via a symlink to this
           directory. If you move these files, it will break the link. So,
           You should probably extract all these files to /usr/src/ to keep
           them safe long term.

  remove   Removes the symlinks from /usr/local/bin, but leaves the source
           files in tact.

For support, or to open issues, create a Github issue.

EOF
exit 0
}

if [ $# -ne 1 ];
    then show_help
fi

case "$1" in
    install)
        check_root
        install_files
        ;;
    remove)
        check_root
        remove_files
        ;;
    help)
        show_help
        ;;
    *)
        show_help
        ;;
esac

create-app.php
<?php

namespace AntAppCreator;

include('classes/AppCreator.class.php');


function show_help() {
?>

SUMMARY
    Creates a PHP-Ant app.

SYNTAX
    create-app [target path]

    You may specify "." as the target path.

EXAMPLES:

    To create an app in the current directory:

      create-app .

    To create an app in a different directory:

      create-app /foo/path/

NOTES:
    - You MUST have already downloaded and installed the ant-signer package,
      and generated your private and public keys with `publish -k`

    - This app will prompt you with a wizard to fill out the required data.

    - SSH access control for your remote git repository is recommended.

<?php
    die();
}

if($argc != 2) show_help();

$appTargetPath = $argv[1];

if($appTargetPath == ".") $appTargetPath = getcwd();
if(file_exists($appTargetPath) == false) die (sprintf("Path not found! (%s)", $appTargetPath));

$tmpDir = '/tmp/_phpant_template/';

$Creator = new AppCreator($appTargetPath);
$Creator->verifyPrerequisites();

$data = $Creator->interview();

$Creator->getTemplate($tmpDir);
$Creator->findAndReplace($data);

$targetDir = $Creator->checkoutTargetRepo($data['TARGETREPO']);

$Creator->writeAppFile($targetDir);
$Creator->publishApp($targetDir);
$Creator->addAppFiles($targetDir);
$Creator->commitApp($targetDir);
$Creator->pushApp($targetDir);

printf("\n APP CREATION COMPLETE.\n");

classes/app.php
<?php

namespace Bugzy
\Timer
;

use PHPAnt\Core\AntApp;
use PHPAnt\Core\AppInterface;

use PHPAnt\Core\CommandInvoker;
use PHPAnt\Core\CommandList;

/**
 * App Name: Timer

 * App Description: Allows users to time their work

 * App Version: 1.0
 * App Action: cli-load-grammar  -> loadTimer
       @ 90
 * App Action: cli-init          -> declareMySelf          @ 50
 * App Action: cli-command       -> processCommand         @ 50
 * App Action: load_loaders      -> Timer
AutoLoader @ 50
 *
 * Actions may be remarked out by changing the ":" to a "#"
 * Example remarked action:
 * App Action# foo-action       -> doSomeWork             @ 50
 *
 * CSS and JS injections:
 * 
 * App Action: header-inject-css -> injectCSS              @ 50
 * App Action: footer-inject-js  -> injectFooterJS         @ 50
 * App Action: header-inject-js  -> injectHeaderJS         @ 50
 */

 /**
 * This app adds the Timer
 and commands into the CLI by adding in
 * the grammar for commands into an array, and returning it up the chain.
 *
 * @package      Bugzy

 * @subpackage   Timer

 * @category     Utility

 * @author       Michael Munger
 <michael@highpoweredhelp.com
>
 */ 


class Timer
 extends AntApp  implements AppInterface  {

    /**
     * Instantiates an instance of the Timer
 class.
     * Example:
     *
     * <code>
     * $appTimer
 = new Timer
();
     * </code>
     *
     * @return void
     * @author Michael Munger
 <michael@highpoweredhelp.com
>
     **/

    function __construct() {
        $this->appName = 'Timer
';
        $this->canReload = true;
        $this->path = __DIR__;

        //requires to use the CommandList to get grammar... and to avoid crashes.
        $this->AppCommands = new CommandList();
        $this->loadCommands();
    }

    /**
     * Callback for the cli-load-grammar action, which adds commands specific to this plugin to the CLI grammar.
     * Example:
     *
     * @return array An array of CLI grammar that will be merged with the rest of the grammar. 
     * @author Michael Munger
 <michael@highpoweredhelp.com
>
     **/

    function loadTimer
() {
        $grammar = [];

        $this->loaded = true;

        /* Example command invoker. Use this for adding CLI commands:

        This code will cause the CLI command `foo bar` to execute the method $this->doSomeWork:

        //Hard way. Use the easy way below. This is here FYI because older apps used this way.
        $callback = 'doSomeWork';
        $criteria = ['is' => ['foo bar' => true]];
        $Command = new CommandInvoker($callback);
        $Command->addCriteria($criteria);
        $this->AppCommands->add($Command);
        
        //Easy way
        $Command = new CommandInvoker('doSomeWork');
        $Command->is('foo bar');
        $this->AppCommands->add($Command);

        //Easy way for "starts with"
        $Command = new CommandInvoker('doSomeWork');
        $Command->startsWith('foo bar');
        $this->AppCommands->add($Command);

        */

        $grammar = array_merge_recursive($grammar, $this->AppCommands->getGrammar());
        
        $results['grammar'] = $grammar;
        $results['success'] = true;
        return $results;
    }

    function Timer
AutoLoader() {
        //REGISTER THE AUTOLOADER! This has to be done first thing! 
        spl_autoload_register(array($this,'loadTimer
Classes'));
        return ['success' => true];
    }

    public function loadTimer
Classes($class) {
        
        //Break apart namespaced requests.

        $buffer = explode("\\",$class);
        $class = end($buffer);

        $baseDir = $this->path;

        $candidate_files = array();

        //Try to grab it from the classes directory.
        $candidate_path = sprintf($baseDir.'/classes/%s.class.php',$class);
        array_push($candidate_files, $candidate_path);

        //Loop through all candidate files, and attempt to load them all in the correct order (FIFO)
        foreach($candidate_files as $dependency) {
            if($this->verbosity > 14) printf("Looking to load %s",$dependency) . PHP_EOL;

            if(file_exists($dependency)) {
                if(is_readable($dependency)) {

                    //Print debug info if verbosity is greater than 9
                    if($this->verbosity > 11) print "Including: " . $dependency . PHP_EOL;

                    //Include the file!
                    include($dependency);
                }
            }
        }
        return ['success' => true];
    }
    
    /**
     * Callback function that prints to the CLI during cli-init to show this plugin has loaded.
     * Example:
     *
     * @return array An associative array declaring the status / success of the operation.
     * @author Michael Munger
 <michael@highpoweredhelp.com>
     **/

    function declareMySelf() {
        if($this->verbosity > 4 && $this->loaded ) print("Timer
 app loaded.\n");

        return ['success' => true];
    }

    function processCommand($args) {
        $cmd = $args['command'];

        return ['success' => true];
    }


}

classes/AppCreator.class.php
<?php

namespace AntAppCreator;

/**
 * {CLASS SUMMARY}
 *
 * Date: 2/21/18
 * Time: 7:05 AM
 * @author Michael Munger <michael@highpoweredhelp.com>
 */

class AppCreator
{
    /**
     * @var string fileBuffer Holds the class data that will be written to file.
     */
    public $fileBuffer = '';
    public $basePath   = '';

    private $appTemplateFileName        = 'ant-app-template.php';
    private $autoLoaderTemplateFileName = 'autoloader-template.php';

    private $repoDir             = '/tmp/_phpant_template/';
    private $repo                = "https://github.com/mjmunger/ant-app-template.git";
    private $appTemplate         = null;
    private $autoLoaderTemplate  = null;

    public function __construct($basePath)
    {
        $this->basePath           = (substr($basePath, -1) == '/' ? rtrim($basePath , '/') : $basePath);
        $this->appTemplate        = $this->repoDir . $this->appTemplateFileName;
        $this->autoLoaderTemplate = $this->repoDir . $this->autoLoaderTemplateFileName;
    }

    public function verifyPrerequisites() {
        //Verify we have a phpant key.
        $privateKeyFile = $home = getenv("HOME") . '/.phpant/ant_id';
        if( file_exists($privateKeyFile) == false ) {
            echo PHP_EOL;
            echo "No private PHPAnt signing key found at: $privateKeyFile. Please run publish -k to generate one." . PHP_EOL;
            echo PHP_EOL;
            die('Cannot continue.');
        }

        return true;
    }

    /**
     * @param $dir Recursively removes a directory.
     * @author Artefactor https://stackoverflow.com/questions/3338123/how-do-i-recursively-delete-a-directory-and-its-entire-contents-files-sub-dir
     */

    private function rrmdir($dir) {
        if (is_dir($dir)) {
            $objects = scandir($dir);
            foreach ($objects as $object) {
                if ($object != "." && $object != "..") {
                    if (is_dir($dir."/".$object))
                        $this->rrmdir($dir."/".$object);
                    else
                        unlink($dir."/".$object);
                }
            }
            rmdir($dir);
        }
    }

    /**
     * Clones the template repository to a local temp directory.
     */
    public function getTemplate() {

        if(file_exists($this->repoDir) && is_dir($this->repoDir)) $this->rrmdir($this->repoDir);

        $cmd    = "git clone $this->repo $this->repoDir 2>&1 ";
        $result = shell_exec($cmd);
    }

    /**
     * Performs substitutions of template fields from the app template, preparing the app to be written to file.
     */

    public function findAndReplace($replacements) {
        $this->fileBuffer = file_get_contents($this->appTemplate);

        foreach($replacements as $find => $replace) {
            $find = sprintf('${%s}', $find);
            $this->fileBuffer = str_replace($find, $replace,$this->fileBuffer);
        }
    }

    /**
     * @param $targetRepo
     * @return string The name of the directory in which to target repo was cloned.
     */

    public function checkoutTargetRepo($targetRepo) {
        $cmd    = "git clone $targetRepo 2>&1 ";
        $result = shell_exec($cmd);
        $targetDir = $this->basePath . '/' . $this->parseTargetDirName($targetRepo);
        return $targetDir;
    }

    public function writeAppFile($targetDirectory) {
        $targetFilePath = $targetDirectory . '/app.php';
        printf("Writing app file to: %s " . PHP_EOL, $targetFilePath);
        $fh = fopen($targetFilePath,'w');
        fwrite($fh,$this->fileBuffer);
        fclose($fh);

        return file_exists($targetFilePath);
    }
    public function publishApp($targetDir) {
        chdir($targetDir);
        echo "Signing and publishing app with your keys...";
        $cmd = 'publish . 2>&1';
        echo "[ OK ]" . PHP_EOL;
        $result = shell_exec($cmd);
    }

    public function addAppFiles($targetDir) {
        chdir($targetDir);
        echo "Adding new files to repository...";
        $cmd = 'git add * 2>&1';
        echo "[ OK ]" . PHP_EOL;
        $result = shell_exec($cmd);
    }

    public function commitApp($targetDir) {
        chdir($targetDir);
        echo "Committing new files to repo...";
        $cmd = 'git commit -a -m\'auto published initial commit\' 2>&1';
        echo "[ OK ]" . PHP_EOL;
        $result = shell_exec($cmd);
    }

    public function pushApp($targetDir) {
        chdir($targetDir);
        echo "Pushing files to repo...";
        $cmd = 'git push 2>&1';
        echo "[ OK ]" . PHP_EOL;
        $result = shell_exec($cmd);
    }

    public function createApp($data, $targetRepo) {
        $targetDir = '/tmp/_phpant_template/';
        $this->verifyPrerequisites();
        $this->getTemplate($targetDir);
        $this->findAndReplace($data);

        $targetDir = $this->checkoutTargetRepo();

        $result = $this->writeAppFile($targetDir);

        if($result == false) throw new Exception("app.php was not created in $targetDir!");

        $this->publishApp($targetDir);
        $this->addAppFiles($targetDir);
        $this->commitApp($targetDir);
        $this->pushApp($targetDir);
    }

    public function parseTargetDirName($repoURI) {
        $buffer = explode('/', $repoURI);
        $gitName = end($buffer);
        $buffer = explode('.', $gitName );
        $targetDir =  $buffer[0];
        return $targetDir;
    }

    /**
     * Asks questions via the command prompt to collate data for app creation.
     */

    public function interview() {

        $data = [];

        $questions = [];
        $questions['TARGETREPO']       = 'What is the git repo URI for this app?';
        $questions['PROJECT']          = 'What is the namespace for the project?';
        $questions['SUBSPACE']         = 'What is the subspace for the project?';
        $questions['CATEGORY']         = 'What is the category for this app?';
        $questions['SYSTEMNAME']       = 'What is the system name for this app? (No spaces, punctuation, etc...)';
        $questions['FRIENDLYNAME']     = 'What is the friendly (human readable) name for this App?';
        $questions['APPDESCRIPTION']   = 'What does this app do?';
        $questions['AUTHORNAME']       = 'What\'s your name?';
        $questions['AUTHOREMAIL']      = 'What\'s your email?';

        foreach($questions as $key => $prompt) {
            printf("$prompt\n");
            $input = trim(fgets(STDIN));
            $data[$key] = $input;
        }

        return $data;
    }

}